no-vary-query.md

```diff
@@ -0,0 +1,264 @@
+
+The proposal augments [RFC 7234](https://httpwg.org/specs/rfc7234.html#rfc.section.4)'s notion of "Constructing Responses from Caches", providing a mechanism to relax the requirement of "The presented effective request URI and that of the stored response match" in well-defined ways. Since everything about the HTTP cache is best-effort, using `No-Vary-Query` will not guarantee additional cache hits. But it should help.
+
+#### … preloading caches
+
+This repository contains specifications for in-memory URL-keyed caches for [prefetch records](https://wicg.github.io/nav-speculation/prefetch.html#document-prefetch-records) and [prerendering browsing contexts](https://wicg.github.io/nav-speculation/prerendering.html#document-prerendering-browsing-contexts-map). This proposal updates the key construction and matching procedure for these caches.
+
+_Note: our [intent](https://github.com/WICG/nav-speculation/issues/170), not yet reflected in the spec, is for these caches to respect at least some of the general `Vary` header semantics._
+
+#### … the cache API
+
+The [Cache API](https://developer.mozilla.org/en-US/docs/Web/API/Cache) already respects the `Vary` header when doing its cache matching, per [special handling in its spec](https://w3c.github.io/ServiceWorker/#request-matches-cached-item). This proposal adds similar special handling for `No-Vary-Query`.
+
+The existing `ignoreVary` option to [`cache.match()`](https://developer.mozilla.org/en-US/docs/Web/API/Cache/match) will _not_ affect the processing of `No-Vary-Query`. `ignoreVary` is useful for increasing the default cache hit rate, by ignoring the requirements of the `Vary` header; having it decrease the default cache hit rate when applied to responses with `No-Vary-Query` headers would be surprising.
+
+The existing `ignoreSearch` option to `cache.match()` will override the effect of `No-Vary-Query`; that is, if `ignoreSearch` is specified, then `No-Vary-Query` is ignored.
+
+We could contemplate adding an `ignoreNoVaryQuery` option to `cache.match()` in the future, but are not yet aware of any use cases for that level of customization.
+
+#### … other browser caches
+
+There are several other URL-keyed caches on the web platform. Examples are:
+
+- The [list of available images](https://html.spec.whatwg.org/#the-list-of-available-images)
+- [Module maps](https://html.spec.whatwg.org/#module-map)
+- [Shared workers](https://html.spec.whatwg.org/#shared-workers-and-the-sharedworker-interface:concept-sharedworkerglobalscope-constructor-url)
+- [Resources preloaded via `<link rel="preload">`](https://html.spec.whatwg.org/#map-of-preloaded-resources)
+
+The proposal does _not_ update cache key computation for these caches. This is because:
+
+- None of them respect the `Vary` header today
+- All of them are keyed by the _request_ URL, and not by the _response_ URL, whereas `No-Vary-Query` is a property provided by the response.
+
+### Interaction with prerendering activation
```

Comment:
It seems like it's probably also worth covering the general case even though it's trivial, i.e., that `window.location` (and everything else that looks at history) reflects the URL that was navigated to (i.e., the URL used to query the cache), not the request/response URL for the fetch that populated the cache.

This applies both to prefetching and the general cache application.

Reply:
I'll expand on the consequences of the URL and history update steps, but I'm not sure I understand your last point.

Reply:
Hmm, I think I see what you mean now. Let me try.

Reply:
Yeah, you figured out what I meant. :)